### ISSUE TITLE
Error in `eps` for certain negative `x`

### ISSUE THREAD

**Issue by dpsanders** [JuliaLang/julia#13447]
This came up while attempting to implement @simonbyrne 's suggestion in #13155 that `eps` should be defined for `BigFloat`s as it is for `Float64`:

```
julia> x = -1.
-1.0

julia> eps(x)
2.220446049250313e-16

julia> nextfloat(x) - x
1.1102230246251565e-16
```

The same holds for any negative `x` that are powers of 2, i.e. where the exponent changes when `nextfloat` is taken.

**Comment by vtjnash** [comment]
I guess it depends on how you define eps?
```
julia> x - prevfloat(x)
2.220446049250313e-16
```
If you check the mantissa, -1.0 is the continuation of the sequence from prevfloat, not nextfloat:
```
julia> hex(reinterpret(UInt64, prevfloat(x)))
"bff0000000000001"

julia> hex(reinterpret(UInt64, x))
"bff0000000000000"

julia> hex(reinterpret(UInt64, nextfloat(x)))
"bfefffffffffffff"
```
while the situation is reversed for positive numbers:
```
julia> x=1.
1.0

julia> hex(reinterpret(UInt64, prevfloat(x)))
"3fefffffffffffff"

julia> hex(reinterpret(UInt64, x))
"3ff0000000000000"

julia> hex(reinterpret(UInt64, nextfloat(x)))
"3ff0000000000001"
```

**Comment by dpsanders** [comment]
Yes, of course you're right. I was being led astray by the documentation:
```
eps(x)

  The distance between x and the next larger representable floating-point
  value of the same DataType as x.
```
which needs a slight change in wording to deal with negative numbers.

**Comment by vtjnash** [comment]
looking at this again, it seems that `nextfloat` is counting in the opposite direction for negative numbers. i'm not really sure whether that is "right" or "wrong" though.

**Comment by dpsanders** [comment]
`nextfloat` finds the adjacent float in the direction towards `+∞`.

I was expecting `eps(a) = nextfloat(a) - a`; this is what does not work.
In this sense, `eps` and `nextfloat` could be said to be at odds.

**Comment by simonbyrne** [comment]
I think this is a definition problem: `eps(x)` should really be the unit in last place of `x`.

We could change the definition to "next largest in magnitude"?

**Comment by JeffreySarnoff** [comment]
I assume nextfloat(x) [prevfloat] is based on nextafter(x,[direction]) which, without the direction, usually means nextafter(x,+Inf).
The IEEE754-2008 standard replaces this with nextUp(x), nextDown(x):
"nextUp(x) is the least floating-point number in the format of x that compares greater than x. If x is
the negative number of least magnitude in x’s format, nextUp(x) is −0. nextUp(±0) is the positive
number of least magnitude in x’s format. nextUp(+∞) is +∞, and nextUp(−∞) is the finite negative
number largest in magnitude. When x is NaN, then the result is according to 6.2. nextUp(x) is
quiet except for sNaNs. The preferred exponent is the least possible. nextDown(x) is −nextUp(−x)."
Perhaps we should either introduce these to replace nextfloat prevfloat or define  nextfloat as nextUp and prevfloat as nextDown.

**Comment by dpsanders** [comment]
I agree that it seems like `eps(x)` should be defined as 1 unit in the last place, which is often called `ulp` according to 

https://en.wikipedia.org/wiki/Unit_in_the_last_place

+1 for "next largest in magnitude".

Though there they comment on these several possible definitions.

`nextup` and `nextdown` sound like reasonable names, but according to the definition you give they seem to be equal to the current `nextfloat` and `prevfloat`.

**Comment by JeffreySarnoff** [comment]
for a numerical analyst's view of ulp, eps and related: http://www.ti3.tuhh.de/paper/rump/RuZiBoMe08.pdf

**Comment by simonbyrne** [comment]
Docs are fixed by #20597.
